File: JAVA/CountSubarraysXOR.java
import java.util.HashMap;

public class CountSubarraysXOR {
    public static int countSubarraysXOR(int[] nums, int k) {
        HashMap<Integer, Integer> xorFreq = new HashMap<>();
        int count = 0, prefixXOR = 0;

        for (int num : nums) {
            prefixXOR ^= num;

            if (prefixXOR == k) {
                count++;
            }

            if (xorFreq.containsKey(prefixXOR ^ k)) {
                count += xorFreq.get(prefixXOR ^ k);
            }

            xorFreq.put(prefixXOR, xorFreq.getOrDefault(prefixXOR, 0) + 1);
        }

        return count;
    }

    public static void main(String[] args) {
        int[] nums1 = {4, 2, 2, 6, 4};
        int k1 = 6;
        System.out.println(countSubarraysXOR(nums1, k1)); // Output: 4

        int[] nums2 = {5, 6, 7, 8, 9};
        int k2 = 5;
        System.out.println(countSubarraysXOR(nums2, k2)); // Output: 2
    }
}

File: Python/CountSubarraysXOR.py
from typing import List
from collections import defaultdict

def count_subarrays_xor(nums: List[int], k: int) -> int:
    xor_freq = defaultdict(int)
    count = 0
    prefix_xor = 0

    for num in nums:
        prefix_xor ^= num

        if prefix_xor == k:
            count += 1

        count += xor_freq[prefix_xor ^ k]
        xor_freq[prefix_xor] += 1

    return count

# Test cases
nums1 = [4, 2, 2, 6, 4]
k1 = 6
print(count_subarrays_xor(nums1, k1))  # Output: 4

nums2 = [5, 6, 7, 8, 9]
k2 = 5
print(count_subarrays_xor(nums2, k2))  # Output: 2
